services/merak-topo/handler/topo_gen.go

```diff
@@ -27,269 +29,424 @@ func GenUUID() string {
 	return strings.Replace(uuidWithHyphen.String(), "-", "", -1)
 }
 
-func Pod_name(num int, prefix string) []string {
-	var pod_name []string
-	var pod_n string
-
-	if prefix == "vswitch" {
-		for i := 0; i < num; i++ {
-			pod_n = prefix + "-" + strconv.FormatInt(int64(i+1), 10)
-			pod_name = append(pod_name, pod_n)
-		}
-	} else {
-		for i := 0; i < num; i++ {
-			pod_n = prefix + "-" + strconv.FormatInt(int64(i), 10)
-			pod_name = append(pod_name, pod_n)
-		}
-	}
-	return pod_name
-}
-
-func Intf_name(dev_num int, prefix string) []string {
-	var intf_name = []string{}
-	var intf_n string = ""
-	if strings.Contains(prefix, "vhost") {
-		intf_n = strings.Split(prefix, "-")[0] + strings.Split(prefix, "-")[1] + "-eth1"
-		intf_name = append(intf_name, intf_n)
-	} else {
-		for i := 0; i < dev_num; i++ {
-			intf_n = strings.Split(prefix, "-")[0] + strings.Split(prefix, "-")[1] + "-eth" + strconv.FormatInt(int64(i+1), 10)
-			intf_name = append(intf_name, intf_n)
-		}
-	}
-	return intf_name
-}
-
-func ip_gen_sub(upper int, k1 int, data_plane_cidr string) []string {
-	var ips_gen = []string{}
+func ip_gen(vhost_idx int, data_plane_cidr string, upper int) string {
 	var ip string = data_plane_cidr
-
 	switch mask := strings.Split(data_plane_cidr, "/")[1]; mask {
 	case "8":
 		//
 	case "24":
 		//
 	default: //16
-		for i := 0; i < upper; i++ {
-			ip = strings.Split(data_plane_cidr, ".")[0] + "." + strings.Split(data_plane_cidr, ".")[1] + "." + strconv.FormatInt(int64(k1), 10) + "." + strconv.FormatInt(int64(i+1), 10) + "/16"
-			ips_gen = append(ips_gen, ip)
-		}
+		i := (vhost_idx) / upper
+		// if (vhost_idx)%upper > 0 {
+		// 	i = i + 1
+		// }
+		ip = strings.Split(data_plane_cidr, ".")[0] + "." + strings.Split(data_plane_cidr, ".")[1] + "." + strconv.FormatInt(int64(i), 10) + "." + strconv.FormatInt(int64((vhost_idx-i*upper)), 10) + "/16"
 	}
 
-	return ips_gen
+	return ip
 }
 
-func Ips_gen(topo_id string, ip_num int, data_plane_cidr string) []string {
-	var count int = 250
-	var ips = []string{}
+func create_vswitches(racks []database.Vnode, init_idx_vs int, ports_per_vswitch int, uid_initial int) (error, []database.Vnode, []database.Vnode) {
+	var vswitches []database.Vnode
+	var racks_attached []database.Vnode
+	num_of_vs := len(racks) / ports_per_vswitch
 
-	t := int(ip_num / count)
+	if len(racks)%ports_per_vswitch > 0 {
+		num_of_vs = num_of_vs + 1
+	}
 
-	if t == 0 {
-		ips = ip_gen_sub(ip_num, 0, data_plane_cidr)
-	} else {
-		ips = ip_gen_sub(count, 0, data_plane_cidr)
-		for iter := 1; iter < t; iter++ {
-			ips = append(ips, ip_gen_sub(count, iter, data_plane_cidr)...)
+	j := init_idx_vs
+	k := ports_per_vswitch
+	for i := 0; i < num_of_vs; i++ {
+		lower_bound := i * k
+		var upper_bound int
+		if i*k+k-1 < len(racks) {
+			upper_bound = lower_bound + k
+		} else {
+			upper_bound = len(racks)
```

Comment:
@zzengcs if possible, we should briefly explain what algorithm you're using here.